llama_index/vector_stores/singlestoredb.py

```diff
@@ -0,0 +1,246 @@
+            )
+        return s2.connect(**self.connection_kwargs)
+    
+    def _create_table(self) -> None:
+        conn = self.connection_pool.connect()
+        cursor = conn.cursor()
+        try:
+            cur = conn.cursor()
+            try:
+                cur.execute(
+                    f"""CREATE TABLE IF NOT EXISTS {self.table_name}
+                    ({self.content_field} TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci,
+                    {self.vector_field} BLOB, {self.metadata_field} JSON);"""
+                )
+            finally:
+                cur.close()
+        finally:
+            conn.close()
+
+    def add(self, nodes: List[BaseNode]) -> List[str]:
+        """Add nodes to index.
+
+        Args
+            nodes: List[BaseNode]: list of nodes with embeddings
+
+        """
+        conn = self.connection_pool.connect()
+        cursor = conn.cursor()
+        try:
+            for node in nodes:
+                embedding = node.get_embedding()
+                metadata = node_to_metadata_dict(
+                    node, remove_text=True, flat_metadata=self.flat_metadata
+                )
+                cursor.execute(
+                        "INSERT INTO {} VALUES (%s, JSON_ARRAY_PACK(%s), %s)".format(
+                            self.table_name
+                        ),
+                        (
+                            node.get_content(metadata_mode=MetadataMode.NONE) or "",
+                            "[{}]".format(",".join(map(str, embedding))),
+                            json.dumps(metadata),
+                        ),
+                    )
+        finally:
+            cursor.close()
+            conn.close()
+        return [node.node_id for node in nodes]
+
+    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
+        """
+        Delete nodes using with ref_doc_id.
+
+        Args:
+            ref_doc_id (str): The doc_id of the document to delete.
+
+        """
+        conn = self.connection_pool.connect()
+        cursor = conn.cursor()
+        try:
+            cursor.execute(f"DELETE FROM {self.table_name} WHERE JSON_EXTRACT_JSON(metadata, 'ref_doc_id') = %s", ('\"' + ref_doc_id + '\"',))
+        finally:
+            cursor.close()
+            conn.close()
+
+    def query(self, query: VectorStoreQuery, filter: Optional[dict] = None) -> VectorStoreQueryResult:
+        """
+        Query index for top k most similar nodes.
+
+        Args:
+            query (VectorStoreQuery): Contains query_embedding and similarity_top_k attributes.
+            filter (Optional[dict]): A dictionary of metadata fields and values to filter by. Defaults to None.
+
+        Returns:
+            VectorStoreQueryResult: Contains nodes, similarities, and ids attributes.
+        """
+        query_embedding = query.query_embedding
+        similarity_top_k = query.similarity_top_k
+        conn = self.connection_pool.connect()
```

Comment:
nit: is there performance concern with making a connection on each request?

Reply:
Good point – since we're using a connection pool here, it's actually keeping the connections alive and I believe it's best practice to connect then close out in each function